src/txrequest.cpp

```diff
@@ -0,0 +1,685 @@
+
+        // Find last entry for this txhash, and extract per-txhash information from it.
+        Iter<ByTxHash> it_last = m_index.get<ByTxHash>().end();
+        // First find the first entry past this txhash.
+        it_last = m_index.get<ByTxHash>().lower_bound(EntryTxHash{gtxid.GetHash(), State::TOO_LARGE, 0});
+        if (it_last != m_index.get<ByTxHash>().begin() && std::prev(it_last)->m_txhash == gtxid.GetHash()) {
+            it_last--;
+        } else {
+            // No entry for this txhash exists yet.
+            it_last = m_index.get<ByTxHash>().end();
+        }
+
+        // Try creating the entry with CANDIDATE_DELAYED state (which will fail due to the uniqueness
+        // of the ByPeer index if a non-CANDIDATE_BEST entry already exists with the same txhash and peer).
+        // Bail out in that case.
+        auto ret = m_index.get<ByPeer>().emplace(gtxid, peer, preferred, reqtime, m_sequence);
+        if (!ret.second) return;
+
+        // Update accounting metadata.
+        ++m_peerinfo[peer].m_total;
+        ++m_sequence;
+    }
+
+    //! Find the GenTxids to request now from peer.
+    std::vector<GenTxid> GetRequestable(uint64_t peer, std::chrono::microseconds now)
+    {
+        // Move time.
+        SetTimePoint(now);
+
+        // Find all CANDIDATE_BEST entries for this peer.
+        std::vector<std::pair<uint64_t, const Entry*>> selected;
+        auto it_peer = m_index.get<ByPeer>().lower_bound(EntryPeer{peer, true, UINT256_ZERO});
+        while (it_peer != m_index.get<ByPeer>().end() && it_peer->m_peer == peer &&
+            it_peer->GetState() == State::CANDIDATE_BEST) {
+            selected.emplace_back(it_peer->m_sequence, &*it_peer);
+            ++it_peer;
+        }
+
+        // Return them, sorted by sequence number.
+        std::sort(selected.begin(), selected.end());
+        std::vector<GenTxid> ret;
+        for (const auto& item : selected) {
+            ret.emplace_back(item.second->m_is_wtxid, item.second->m_txhash);
+        }
+        return ret;
+    }
+
+    void RequestedTx(uint64_t peer, const GenTxid& gtxid, std::chrono::microseconds expiry)
+    {
+        auto it = m_index.get<ByPeer>().find(EntryPeer{peer, true, gtxid.GetHash()});
+        // RequestedTx can only be called on CANDIDATE_BEST entries (this is implied by its condition that it can
+        // only be called on GenTxids returned by GetRequestable (and only AlreadyHave and RequestedTx can be called
+        // in between, which preserve the state of other GenTxids).
+        assert(it != m_index.get<ByPeer>().end());
+        assert(it->GetState() == State::CANDIDATE_BEST);
+        Modify<ByPeer>(it, [expiry](Entry& entry) {
+            entry.SetState(State::REQUESTED);
+            entry.m_time = expiry;
+        });
+
+        // Update the per-txhash data (of the last Entry for this txhash) to reflect that new ones are no longer
+        // eligible for the "first" marker.
```

Comment:
There is no longer a "first" maker for the entires, is there?

Reply:
Gone.